blog/2024-02-02-migrating-machines-to-xstate-v5/index.mdx

```diff
@@ -0,0 +1,651 @@
+    };
+    events: {} as {
+    };
+  },
+})
+```
+
+That still works but but it’s even easier to just cast the whole types object all at once.
+
+```ts
+const machine = setup({
+  types: {} as {
+    context: {};
+    events: {};
+  },
+});
+```
+
+:::warning
+
+When chaining `setup().createMachine()` in v5, `types` can only be defined on the object passed to `setup()` and not on the config object passed to `.createMachine()`. Example: `setup({ types: {} }).createMachine({})`. A chained call to `createMachine()` only accepts the one argument, the machine config.
+
+:::
+
+### Remove typegen
+
+If you were previously using typegen in v4, you can also delete `tsTypes` as there is no more typegen in v5. Typing events in actions and guards can be done using type narrowing in those implementations. However, the following sections will show you how you can type a new `params` argument passed to those implementations.
+
+## 3. Convert action and guard strings to parameterized objects
+
+We must now [convert any named action or guard strings to action objects](/docs/migration#use-params-to-pass-params-to-actions--guards) which allows us to define an explicit `params` object (or primitive value) that will be received at runtime by our action implementations. There are two ways to do this:
+
+We can define `params` as a [static value](/docs/actions#action-objects)
+
+```ts
+actions: [
+  {
+    // Action type
+    type: 'track',
+    // Action params
+    params: { response: 'good' },
+  },
+],
+```
+
+We can also use [dynamic action parameters](/docs/actions#dynamic-action-parameters), a function that receives `context` and/or `event`, allowing you to map those values to the `params` object.
+
+```ts
+entry: [
+  {
+    type: 'logInitialRating',
+    params: ({ context }) => ({
+      initialRating: context.initialRating,
+    }),
+  },
+];
+```
+
+Here is a more complete example of converting action strings to parameterized actions:
+
+<Tabs>
+<TabItem value="v5" label="XState v5">
+
+```ts
+import { setup } from 'xstate';
+
+const machine = setup({
+  types: {
+    events: {} as {
+      type: 'next';
+      prop1: string;
+      prop2: number;
+      prop3: boolean;
+    },
+  },
+  /* more setup */
+}).createMachine({
+  on: {
+    next: {
+      target: 'first',
+      actions: [
+        {
+          guard: {
+            type: 'is this ready',
+            params: ({ event }) => ({ ready: event.ready }),
+          },
+          type: 'doThis',
+          // Later, the action implementation function will be
+          // passed this string value in the 2nd arg, params.
+          params: ({ event }) => ({ prop1: event.prop1 }),
+        },
+        {
+          type: 'doThat',
+          // Later, the action implementation function will be
+          // passed this number value in the 2nd arg, params.
+          params: ({ event }) => ({ prop2: event.prop2 }),
+        },
+      ],
+    },
+  },
+  states: {
+    first: {
+      entry: {
```

Comment:
this totally works but the downside is that within `entry` and `exit` we can't  narrow `event` automatically, so parameterized actions don't give us a lot here if we are after converting `event` to `params` for the improved type experience. It might still be a good pattern to have everything to use the same pattern etc but the biggest benefits of `params` are seen when used with transition actions

Reply:
Thanks for the clarification. I'll definitely mention the difference between `entry` and `exit` actions and transition actions here.